### PR TITLE
Add SUPG to INS

### PR DESCRIPTION
**Pull request by lindsayad**
This adds consistent (omitting terms with second derivatives) streamline upwinding for incompressible Navier-Stokes. SUPG is a method that modifies the test function, essentially giving more weight to the upwind element. An optimal factor (alpha) for the upwinding is chosen such that at least for single component, one dimension scalar advection, nearly exact solutions at the nodes are obtained.

SUPG allows higher Reynolds and grid Peclet numbers to be simulated. An example of a simulation that will not converge at steady-state with standard Galerkin weighting, but does converge with PG weighting is shown below with comparison to an OpenFOAM simulation. The Reynolds number for this problem is 1000; the grid Peclet number is 25.

### moose

moose_supg_lid_driven

### OpenFOAM

foam_lid_driven

Some other fun examples of the importance of consistent SUPG weighting (e.g. application of the PG perturbation to the test function to source terms and time derivatives) are shown below:

### Importance of consistent weighting of source terms

The "No SUPG" label here means that SUPG is applied only to the advection operator

source_supg_effect_hill

### Importance of consistent weighting of time derivative

transient_supg_effect

The latter two example problems are taken from Huges et. al. Computer Methods in Applied Mechanics and Engineering 32 (1982) 199-259.

Addresses #9643

### CODE REVIEW

**Comment by lindsayad** [comment]
Interest in this @jwpeterson ?

**Comment by jwpeterson** [comment]
> Interest in this @jwpeterson?

Yes, let's look at this during the Tiger Team this week?

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+  }
+  Real PG_test = alpha * _current_elem->hmax() / 2. * U / U.norm() * _grad_test[_i][_qp];
+
+  Real convective_part = _rho[_qp] * U * _grad_u[_qp] * PG_test;
+
+  // Note that integrating pressure by parts would imply applying a second
+  // derivative to the test function, which we are not equipped to do (?)
+  Real pressure_part = 0;
+  if (!_integrate_p_by_parts)
+    pressure_part = _grad_p[_qp](_component) * PG_test;
```
Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: Per our discussion, I think the stabilization terms should involve grad(p) regardless of whether or not we integrate by parts.

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+  // For reasonable (linear) element shapes the contribution of PG_test * viscous
+  // term is small and can be neglected. This is not generally the case for higher
+  // order elements (and note that we commonly use quadratic shape functions for
+  // velocity). However, including this term would again imply applying a second
+  // derivative either to the test function or to the variable u
+  // Real viscous_part = computeQpResidualViscousPart();
+
+  Real body_force_part = -_rho[_qp] * _gravity(_component) * PG_test;
+
+  return convective_part + pressure_part + /*viscous_part +*/ body_force_part;
```
Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: As a longer term design goal, I think the "unstabilized" `INSMomentum` Kernel and the `INSMomentumSUPG` kernels should share a common base class which provides the strong (and weak) forms of the residual in a single location. That way, if new terms are added to the equation, they can be added in a single place rather than two different files. Code would also be able to be shared, since some parts of the strong and weak residual are the same (the parts that are not integrated by parts).

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+Real
+INSMomentumSUPG::computeQpResidual()
+{
+  RealVectorValue U(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
+  Real alpha;
+  if (_mu[_qp] < std::numeric_limits<double>::epsilon())
+    alpha = 1;
+  else if (U.norm() < std::numeric_limits<double>::epsilon())
+  {
+    alpha = 0;
```
Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: Setting `alpha=0` is not necessary if you are just going to return.

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C (hunk 2 above)
Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: Not necessary to fix for this PR, but I'd like to know the tradeoffs involved in neglecting the viscous terms of the residual. Specifically, we should have some MMS tests to verify the convergence rates are still optimal when we make this error in consistency.

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    Real alpha;
+    Real d_alpha_d_vel_component;
+    if (_mu[_qp] < std::numeric_limits<double>::epsilon())
+    {
+      alpha = 1;
+      d_alpha_d_vel_component = 0;
+    }
+    else if (U.norm() < std::numeric_limits<double>::epsilon())
+    {
+      alpha = 0;
```
Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: redundant

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
```
Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: redundant

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+          d_grid_Pe_d_vel_component;
+    }
+    Real PG_test = alpha * _current_elem->hmax() / 2. * U / U.norm() * _grad_test[_i][_qp];
+    RealVectorValue d_U_d_vel_component(0, 0, 0);
+    d_U_d_vel_component(0) = _phi[_j][_qp];
+    Real d_PG_test_d_vel_component =
+        _current_elem->hmax() / 2. * _grad_test[_i][_qp] *
+        (d_alpha_d_vel_component * U / U.norm() +
+         alpha * (d_U_d_vel_component / U.norm() -
+                  U / (U.norm() * U.norm() * U.norm()) * _u_vel[_qp] * _phi[_j][_qp]));
```
Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: The only difference between this line and the other components is `_u_vel` vs. `_v_vel` vs. `_w_vel`, right? Would it be possible to reduce some of the code duplication by mapping `jvar` into a `compnent` and using `U(component)` instead?

File: modules/navier_stokes/tests/ins/jacobian_test/jacobian_supg_test.i [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  ymax = 1.1
+  zmax = 1.1
+  nx = 1
+  ny = 1
+  nz = 1
+  elem_type = HEX8
+[]
+
+[Variables]
+  [./vel_x]
```
Comment (jwpeterson) [inline modules/navier_stokes/tests/ins/jacobian_test/jacobian_supg_test.i]: This test uses equal-order interpolation for velocity and pressure, so it is not LBB stable. Although it might work for the specific grid used in this test, you would potentially encounter issues if you try to refine the mesh, etc. I'd rather not have input files with unstable formulations in the module, because people create new simulations and tests by copying existing ones...

File: modules/navier_stokes/tests/ins/lid_driven/lid_driven_supg.cmp [hunk 8]
```diff
@@ -0,0 +1,10 @@
+COORDINATES absolute 1.e-6    # min separation not calculated
+
+TIME STEPS relative 1.e-6 floor 0.0     # min:               0 @ t1 max:             2.5 @ t6
+
+# No GLOBAL VARIABLES
+
+NODAL VARIABLES relative 1.e-6 floor 1.e-12
+	p      absolute 5.e-4
```
Comment (jwpeterson) [inline modules/navier_stokes/tests/ins/lid_driven/lid_driven_supg.cmp]: Seems like a very loose tolerance for pressure? Is this just copied from the other lid-driven custom comparison file?

**Review by jwpeterson — CHANGES_REQUESTED** [review]
* Fix redundant alpha=0 settings
* Look into reducing code duplication in Jacobian calculations.
* Fix LBB-unstable Jacobian test.
* Double-check loose pressure tolerance on new lid-driven cavity test.

File: modules/navier_stokes/src/kernels/INSMomentumSUPG.C (hunk 2 above)
Comment (lindsayad) [inline modules/navier_stokes/src/kernels/INSMomentumSUPG.C]: Started working on PSPG (will be a different PR) and have created a base class like you suggested

**Review by jwpeterson — APPROVED** [review]
